### PR TITLE
Clockwork golem megafauna

### PR DESCRIPTION
**Pull request by Comxy**
## About The Pull Request
Adds a new clockwork golem boss to lavaland, spawning on the dead ravtar surface.
Golem has two stages, once it dies it spawns a smaller version that tries to murder you very quickly.
First stage has four abilities, smoke, oil ball, ruby blast and spawn spider.
Smoke and spawn spider speaks for themselves.
Oil ball shoots a ball that spawns slippery oil puddles, bounces of the walls and knocks the target down.
Ruby blast shoots a ball that explodes on impact.
The golem also has the passive ability to do damage to your heart directly with his drill.

Reward for killing first stage is a KA modkit that drops oil on the ground when shot.
Also adds a skin for the modkit.

Reward for killing second stage is a book that teaches oil jaunt.
Oil jaunt works only when you jump from a tile with oil to another tile with oil.

Reward for killing second stage with a crusher gives a clockwork rocket trophy.
Clockwork trophy increases range, a bit of cooldown and speed of the crusher bolt.

## Why It's Good For The Game
It is a cool boss and is cool for the lore, that there are still active clockwork remains.
This spawning at the clockwork remains of ravtar sets a great environment and adds to the boss, making it not seem like an out of place spawn like other creatures on lavaland.
The abilities are balanced, well thought out and tested.
The code is very good, since it is coded using references and it all works flawlessly.
This pr gives some cool extra content for the miners to explore.
promclock1
promclock2
Finally something different than ash drake, colossus, legion or bubblegum.

## Changelog

:cl:
add: Added a clockwork lavaland boss with a spell and a trophy as loot.
soundadd: Added sprites of abilities and a spell, as well as sprites for mobs.
imageadd: Added cool clockwork sounds.
/:cl:

### CODE REVIEW

**Comment by SmArtKar** [comment]
Loot seems extremely underwhelming tbh. Mining loot should be useful for, well, mining - and theres not a lot of oil on lavaland(theres none). Same for the crusher trophy - very meh.

File: code/game/objects/effects/decals/cleanable.dm [hunk 1]
```diff
@@ -110,6 +110,12 @@
 	else
 		return 0
 
+/obj/effect/decal/cleanable/proc/can_oiljaunt_in()
+	if(blood_state == BLOOD_STATE_OIL)
+		return bloodiness
+	else
+		return 0
```
Comment (Fikou) [inline code/game/objects/effects/decals/cleanable.dm:117]: use TRUE/FALSE defines

File: code/game/objects/items/granters.dm [hunk 2]
```diff
@@ -333,6 +333,18 @@
...
+	spell = /obj/effect/proc_holder/spell/pointed/oil_jaunt
+	spellname = "oiljaunt"
+	icon_state = "bookoiljaunt"
+	desc = "This book is sad, it is crying oil."
+	remarks = list("It is written in an ancient language, damn guess this is a waste of time.", "So I guess I am still trying to read this, I kind of think I know what it says.", "Everything begins with or without energy", "Become like oil, formless.", "Now you put oil in a book, it becomes the book.", "Now you put oil in yourself.", "Your time is running faster.")
+
+/obj/item/book/granter/spell/oiljaunt/recoil(mob/living/user)
+	..()
+	to_chat(user,span_warning("After trying to read the book, you realize it is too heavy to hold."))
+	user.Knockdown(30)
```
Comment (Fikou) [inline code/game/objects/items/granters.dm:346]: use SECONDS defines

File: code/game/sound.dm [hunk 3]
```diff
@@ -222,6 +222,8 @@ distance_multiplier - Can be used to multiply the distance at which the sound is
 				soundin = pick('sound/effects/rustle1.ogg','sound/effects/rustle2.ogg','sound/effects/rustle3.ogg','sound/effects/rustle4.ogg','sound/effects/rustle5.ogg')
 			if (SFX_BODYFALL)
 				soundin = pick('sound/effects/bodyfall1.ogg','sound/effects/bodyfall2.ogg','sound/effects/bodyfall3.ogg','sound/effects/bodyfall4.ogg')
+			if (SFX_CLOCKFALL)
+				soundin = 'sound/creatures/clockwork_golem_dead.ogg'
```
Comment (Fikou) [inline code/game/sound.dm:226]: theres no reason to use the string defines for just 1 file, you can use that file directly

File: code/modules/mining/equipment/kinetic_crusher.dm [hunk 4]
```diff
@@ -441,3 +441,26 @@
 		chaser.monster_damage_boost = FALSE // Weaker cuz no cooldown
 		chaser.damage = 20
 		log_combat(user, target, "fired a chaser at", src)
+
+//clockwork golem
+/obj/item/crusher_trophy/clockwork_rocket
```
Comment (Fikou) [inline code/modules/mining/equipment/kinetic_crusher.dm:446]: this trophy doesnt give you a description of what it does

File: code/modules/mob/living/simple_animal/hostile/megafauna/clockwork_golem.dm [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+	pixel_x = -16
+	gps_name = "Clockwork Signal"
+	del_on_death = TRUE
+	deathmessage = "cracks, beaking into multiple pieces."
+	deathsound = SFX_CLOCKFALL
+	footstep_type = FOOTSTEP_MOB_HEAVY
+
+/mob/living/simple_animal/hostile/megafauna/clockwork_golem/Initialize(mapload)
+	ADD_TRAIT(src, TRAIT_NO_FLOATING_ANIM, INNATE_TRAIT)
+	. = ..()
```
Comment (Fikou) [inline code/modules/mob/living/simple_animal/hostile/megafauna/clockwork_golem.dm:34]: return ..() at the end of procs

File: code/modules/mob/living/simple_animal/hostile/megafauna/clockwork_golem.dm [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+
+/mob/living/simple_animal/hostile/megafauna/clockwork_golem/death(gibbed)
+	var/datum/effect_system/fluid_spread/smoke/smoke = new
+	smoke.set_up(2, location = loc)
+	smoke.start()
+	. = ..()
```
Comment (Fikou) [inline code/modules/mob/living/simple_animal/hostile/megafauna/clockwork_golem.dm:40]: and here

File: code/modules/mob/living/simple_animal/hostile/megafauna/clockwork_golem.dm [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+/mob/living/simple_animal/hostile/megafauna/clockwork_golem/complete/Destroy()
+	QDEL_NULL(ruby_blast)
+	QDEL_NULL(release_smoke)
+	QDEL_NULL(oil_ball)
+	QDEL_NULL(summon_spider)
+	return ..()
+
+/mob/living/simple_animal/hostile/megafauna/clockwork_golem/complete/death()
+	new /mob/living/simple_animal/hostile/megafauna/clockwork_golem/broken(loc)
+	. = ..()
```
Comment (Fikou) [inline code/modules/mob/living/simple_animal/hostile/megafauna/clockwork_golem.dm:91]: and here

File: code/modules/mob/living/simple_animal/hostile/megafauna/clockwork_golem.dm [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+	new /obj/effect/decal/cleanable/oil/slippery(get_turf(loc))
+	. = ..()
+
+/obj/projectile/bullet/oil_ball/on_hit(atom/target, blocked = FALSE)
+	new /obj/effect/decal/cleanable/oil/slippery(get_turf(loc))
+	. = ..()
+
+/obj/projectile/bullet/oil_ball/on_range()
+	new /obj/effect/decal/cleanable/oil/slippery(get_turf(loc))
+	. = ..()
```
Comment (Fikou) [inline code/modules/mob/living/simple_animal/hostile/megafauna/clockwork_golem.dm:168]: and on these

File: code/modules/spells/spell_types/oil_jaunt.dm [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+	var/turf/mobloc = get_turf(user)
+	new jaunt_in_type(mobloc, user.dir)
+	var/obj/effect/dummy/phased_mob/oil_jaunt/holder = new /obj/effect/dummy/phased_mob/oil_jaunt(mobloc)
+	user.forceMove(holder)
+	ADD_TRAIT(user, TRAIT_IMMOBILIZED, type)
+	sleep(jaunt_time)
+	play_sound("exit", user)
+	new jaunt_out_type(target_loc, user.dir)
+	holder.forceMove(target_loc)
+	sleep(5)
```
Comment (Fikou) [inline code/modules/spells/spell_types/oil_jaunt.dm:56]: use SECONDS defines

File: code/modules/spells/spell_types/oil_jaunt.dm [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+	mobloc = get_turf(user.loc)
+	holder.reappearing = 1
```
Comment (Fikou) [inline code/modules/spells/spell_types/oil_jaunt.dm:58]: use TRUE/FALSE defines

File: code/game/objects/effects/decals/cleanable.dm (hunk 1 above)
Comment (Fikou) [reply code/game/objects/effects/decals/cleanable.dm:117]: you also don't need that else, you already returned

**Comment by Comxy** [comment]
Decreased organ damage to 10. With at least 10u of oil, oil can be thrown to the ground making an oil puddle to balance oil jaunt to be useful in creative situation. Crusher trophy now also does 10 bonus damage on detonation to make it more useful.

File: code/modules/mining/equipment/kinetic_crusher.dm [hunk 11]
```diff
@@ -442,13 +442,13 @@
 		chaser.damage = 20
 		log_combat(user, target, "fired a chaser at", src)
 
-//clockwork golem
+//clockwork golem, increases detonation damage, market range/speed and decreases charge time.
 /obj/item/crusher_trophy/clockwork_rocket
```
Comment (Fikou) [inline code/modules/mining/equipment/kinetic_crusher.dm:446]: i meant as in. the proc that tells the players on examine what it does.

**Comment by Mey-Ha-Zah** [comment]
Okay, as a Miner main, here's my two cents. Organ damage to a critical organ is juts a fat middle finger to any miner who is able to soak the hits (Crusher miners for instance) Now if it was just de-limbing an arm or a leg, thats more understandable, we can play around that. Hell a Miner is never truly a miner until he's lost both arms in a mining accident and lived to tell about it.

As for the Oil Jaunt and attack, thats another thing altogether from my experience. Miner's only non-slip is for ice... and oil is a hard slip if I remember correctly, and any major slowdown is almost a death sentence in a megafauna fight.

**Comment by MetalClone** [comment]
I like the crusher trophy, the loot is fairly unique and not aiming to be anything overpowered. The fight on paper seems fairly hard though, can we get a video or two of fights with the boss?

**Comment by zxaber** [comment]
Oil jaunt would be cool if paired with a pka mod or crusher trophey that drops (non-slip) oil when hitting a mob. You'd end up with a combo where after some time in a fight, you can jaunt around attacks.

**Comment by Cheshify** [comment]
Oh boy let's give miners a DISCOUNT BLOOD CRAWL this cannot go wrong. Sorry but while it's a cool boss this gamer loot is not great.

**Comment by Shadowflame909** [comment]
> Oh boy let's give miners a DISCOUNT BLOOD CRAWL this cannot go wrong. Sorry but while it's a cool boss this gamer loot is not great.

The only place oil puddles spawns on station is outside of robotics

But yeah this damage type is pretty crazy and seems like it'd require S tier powergaming ala ling fleshmend to beat. Its why no bosses do cloning damage.

**Comment by Bluedino1025** [comment]
maybe instead of the oil crawl (bad for miners), you should make it where it drops its drill and if you pick it up it'll be faster than the sonic jackhammer on and off station but you can't take it off or add the oil ball to the spellbook that way you can jaunt and throw oil. Another choice would be to have the spellbook teach a weakened ruby blast that only works on lavaland but can be used for mining and for fighting.

**Review by Maurukas — APPROVED** [review]
Mapping is good.

I could see oilcrawl potentially becoming a problem, as this ruin spawns fairly regularly, but since there's very few mapped oil puddles (1-2 on most maps), a miner is going to mostly be limited to ones they place.

Still cool and lore-friendly enough that I want to see it though.

File: code/datums/actions/mobs/release_smoke.dm [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+/datum/action/cooldown/mob_cooldown/release_smoke/Activate(atom/target_atom)
+	playsound(owner, smoke_sound, 100)
+	StartCooldown(10 SECONDS)
+	create_smoke(target_atom)
+	StartCooldown()
+
+/datum/action/cooldown/mob_cooldown/release_smoke/proc/create_smoke(atom/target)
+	SLEEP_CHECK_DEATH(smoke_delay, src)
+	var/turf/smoke_location = null
+	smoke_location = get_turf(owner)
```
Comment (Kylerace) [inline code/datums/actions/mobs/release_smoke.dm:24]:
```suggestion
	var/turf/smoke_location = get_turf(owner)
```

File: code/datums/actions/mobs/summon_minion.dm [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+/datum/action/cooldown/mob_cooldown/summon_minion/Activate(atom/target_atom)
+	playsound(owner, summon_sound, 100)
+	StartCooldown(10 SECONDS)
+	create_minion(target_atom)
+	StartCooldown()
+
+/datum/action/cooldown/mob_cooldown/summon_minion/proc/create_minion(atom/target)
+	SLEEP_CHECK_DEATH(summon_delay, src)
+	var/turf/summon_location = null
+	summon_location = get_turf(owner)
```
Comment (Kylerace) [inline code/datums/actions/mobs/summon_minion.dm:24]:
```suggestion
	var/turf/summon_location = get_turf(owner)
```

File: code/modules/reagents/chemistry/reagents/other_reagents.dm [hunk 14]
```diff
@@ -1800,6 +1800,17 @@
 	chemical_flags = REAGENT_CAN_BE_SYNTHESIZED
 	addiction_types = null
 
+/datum/reagent/fuel/oil/expose_turf(turf/exposed_turf, reac_volume)
+	. = ..()
+	if(!istype(exposed_turf) || isspaceturf(exposed_turf)) // Is the turf valid
```
Comment (Kylerace) [inline code/modules/reagents/chemistry/reagents/other_reagents.dm:1805]: what about openspace turfs

**Review by Kylerace — CHANGES_REQUESTED** [review]

File: code/modules/reagents/chemistry/reagents/other_reagents.dm (hunk 14 above)
Comment (Comxy) [reply code/modules/reagents/chemistry/reagents/other_reagents.dm:1805]: I am pretty sure it isn't possible to drop in open space.

**Comment by ShizCalev** [comment]
BEGONE BACK TO HELL FOUL BEAST. I KILLED THEE.

**Review by Twaticus — APPROVED** [review]

File: code/modules/mining/equipment/kinetic_crusher.dm [hunk 15]
```diff
@@ -441,3 +441,31 @@
...
+	denied_type = /obj/item/crusher_trophy/clockwork_rocket
+	bonus_value = 10
+
+/obj/item/crusher_trophy/clockwork_rocket/effect_desc()
+	return "mark detonation to do <b>[bonus_value]</b> more damage and decreases the crushers charge time by <b>[bonus_value * 0.1]</b>, while increasing the destabilizer shots range by <b>3</b> tiles and increasing its speed by <b>0.8</b> times"
+
+/obj/item/crusher_trophy/clockwork_rocket/on_projectile_fire(obj/projectile/destabilizer/marker, mob/living/user)
+		marker.name = "extended [marker.name]"
+		marker.speed *= 0.8
+		marker.range += 3
```
Comment (tralezab) [inline code/modules/mining/equipment/kinetic_crusher.dm:459]: double indented

File: code/modules/mob/living/simple_animal/hostile/megafauna/clockwork_golem.dm [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+
+When the clockwork golem dies, it leaves behind a spell book or trophy for your crusher:
+A. The spell book gives oil jaunt that activates when you are targeting and standing on a tile with oil.
+B. The trophy increases range, speed and cooldown of the bolt.
+
+Difficulty: Medium
+
+*/
+
+/mob/living/simple_animal/hostile/megafauna/clockwork_golem
```
Comment (tralezab) [inline code/modules/mob/living/simple_animal/hostile/megafauna/clockwork_golem.dm:20]: remove this space, it makes the autodoc not work :^(

File: code/modules/mob/living/simple_animal/hostile/megafauna/clockwork_golem.dm [hunk 17]
```diff
@@ -0,0 +1,168 @@
+/*
```
Comment (tralezab) [inline code/modules/mob/living/simple_animal/hostile/megafauna/clockwork_golem.dm:1]: turn this into autodoc >:)

**Review by tralezab — CHANGES_REQUESTED** [review]
the nitpicks

**Comment by Comxy** [comment]
I don't feel like leaving this open for too long so I can move on from this pr. Might work on something else in the future I'll see.

**Comment by Shadowflame909** [comment]
rip oil jaunt

**Pull request closed without merging**